node/cn/snap/handler.go

```diff
@@ -0,0 +1,524 @@
+	maxCodeLookups = 1024
+
+	// stateLookupSlack defines the ratio by how much a state response can exceed
+	// the requested limit in order to try and avoid breaking up contracts into
+	// multiple packages and proving them.
+	stateLookupSlack = 0.1
+
+	// maxTrieNodeLookups is the maximum number of state trie nodes to serve. This
+	// number is there to limit the number of disk lookups.
+	maxTrieNodeLookups = 1024
+
+	// maxTrieNodeTimeSpent is the maximum time we should spend on looking up trie nodes.
+	// If we spend too much time, then it's a fairly high chance of timing out
+	// at the remote side, which means all the work is in vain.
+	maxTrieNodeTimeSpent = 5 * time.Second
+)
+
+// Handler is a callback to invoke from an outside runner after the boilerplate
+// exchanges have passed.
+type Handler func(peer *Peer) error
+
+type SnapshotReader interface {
+	StateCache() state.Database
+	Snapshots() *snapshot.Tree
+	ContractCode(hash common.Hash) ([]byte, error)
+	ContractCodeWithPrefix(hash common.Hash) ([]byte, error)
+}
+
+type SnapshotDownloader interface {
+	// DeliverSnapPacket is invoked from a peer's message handler when it transmits a
+	// data packet for the local node to consume.
+	DeliverSnapPacket(peer *Peer, packet Packet) error
+}
+
+// Handle is the callback invoked to manage the life cycle of a `snap` peer.
+// When this function terminates, the peer is disconnected.
+func Handle(reader SnapshotReader, downloader SnapshotDownloader, peer *Peer) error {
+	for {
+		if err := HandleMessage(reader, downloader, peer); err != nil {
+			peer.Log().Debug("Message handling failed in `snap`", "err", err)
+			return err
+		}
+	}
+}
+
+// HandleMessage is invoked whenever an inbound message is received from a
+// remote peer on the `snap` protocol. The remote connection is torn down upon
+// returning any error.
+func HandleMessage(reader SnapshotReader, downloader SnapshotDownloader, peer *Peer) error {
+	// Read the next message from the remote peer, and ensure it's fully consumed
+	msg, err := peer.rw.ReadMsg()
+	if err != nil {
+		return err
+	}
+	if msg.Size > maxMessageSize {
+		return fmt.Errorf("%w: %v > %v", errMsgTooLarge, msg.Size, maxMessageSize)
+	}
+	defer msg.Discard()
+	start := time.Now()
+	// TODO-Klaytn-SnapSync add metric to track the amount of time it takes to serve the request and run the manager
+	// Handle the message depending on its contents
+	switch {
+	case msg.Code == GetAccountRangeMsg:
+		// Decode the account retrieval request
+		var req GetAccountRangePacket
+		if err := msg.Decode(&req); err != nil {
+			return fmt.Errorf("%w: message %v: %v", errDecode, msg, err)
+		}
+		// Service the request, potentially returning nothing in case of errors
+		accounts, proofs := ServiceGetAccountRangeQuery(reader, &req)
+
+		// Send back anything accumulated (or empty in case of errors)
+		return p2p.Send(peer.rw, AccountRangeMsg, &AccountRangePacket{
+			ID:       req.ID,
+			Accounts: accounts,
```

Comment:
`accounts` can be nil. Is it okay?

Reply:
Is there any test case to cover this case?

Reply:
I have added testcases, but encoding/decoding `nil` accounts were okay. Am I missing something? Please take another look.

Reply:
https://github.com/klaytn/klaytn/pull/1473/files#diff-7598d07f999ff8aea6077cbbbf66d0557e29530315933b1a03ba84c8111b75c6R46

If you are using testDownloader, DeliverSnapPacket() does nothing.

Reply:
@jeongkyun-oh I added my test case with the name of handler2_test.go